Fix segfault on creating non-introspected object

wotzlaff opened the pull request:
After some days of digging I came to the conclusion that `GetClassTemplateFromGI` should not be used in `WrapperFromGObject`. I'd propose to use `GetClassTemplate` instead.
At least this solves my issue #179 and doesn't break any tests. Let me know if you see other problems.

Going further: I am not quite sure where we need `GetClassTemplateFromGI` in the first place. Moreover, `GetClassTemplate` might not even need the `gi_info` argument?

romgrk commented on src/gobject.cc:
```diff
@@ -427,18 +427,20 @@ static MaybeLocal<FunctionTemplate> NewClassTemplate (GIBaseInfo *info, GType gt
-    if (parent_info) {
-        auto parent_tpl = GetClassTemplateFromGI ((GIBaseInfo *) parent_info);
+    GType parent_type = g_type_parent(gtype);
+    if (parent_type == 0) {
```
There is probably some constant (`G_TYPE_NONE` maybe?) to use here instead of 0. Zero doesn't carry any semantic value, as opposed to a named constant.

romgrk commented on src/gobject.cc:
```diff
@@ -456,19 +458,21 @@ static MaybeLocal<FunctionTemplate> GetClassTemplate(GIBaseInfo *gi_info, GType
         gi_info = g_irepository_find_by_gtype(NULL, gtype);
 
-    assert_printf (gi_info != NULL, "Missing GIR info for: %s\n", g_type_name (gtype));
+    // assert_printf (gi_info != NULL, "Missing GIR info for: %s\n", g_type_name (gtype));
```
Either remove or uncomment.

wotzlaff commented:
Thanks for your review. You're absolutely right with your short code comments. I fixed it.

By doing this I found that `G_TYPE_NONE == 1` is not the critical type but `G_TYPE_INVALID == 0`. So maybe the checks for `G_TYPE_NONE` in the remaining code should be overthought.

About the travis test: It used to fail before my changes. I didn't change anything with boxed types, so hopefully I didn't introduce the failure.

About the GLib/GIR question: I think, I'm still using GIR after my changes. But actually, I am not pretty sure about it. Also, the implications are very unclear to me. In the last commit, I also removed `GetClassTemplateFromGI`, which could make it even worse... I think the issue boils down to the question whether we're able to construct the class solely from the given `GType`. This seems to be possible: `g_irepository_find_by_gtype` gives us at runtime something for introspected types. This info can then be used for whatever we need (basically in the same way as with pure/static GIR).
After all, I haven't fully understood the whole GIR infrastructure. So maybe everything I am writing here is fully nonsense.
I just saw that `NewClassTemplate` is also not using the `GIBaseInfo` anymore. I will strip that argument out, too.

I added a test for the creation of nongi-objects using GStreamer. I think it's not really a goo way because then GStreamer is needed as a dependency for running the tests which might not be installed in every environment. But I do not see any other example from my (low) experience.

wotzlaff commented:
After the comment above, I stripped down `NewClassTemplate` a bit. Hopefully these update do not break anything :)

romgrk commented on src/gobject.cc:
```diff
@@ -453,45 +452,29 @@ static MaybeLocal<FunctionTemplate> GetClassTemplate(GIBaseInfo *gi_info, GType
-
-MaybeLocal<Function> MakeClass(GIBaseInfo *info) {
-    auto tpl = GetClassTemplateFromGI (info);
+    auto tpl = GetClassTemplate(gtype);
```
Minor formatting issue here, I'd add newlines like illustrated to make it less cramed together, it's more readable.

![Screenshot from 2020-05-26 22-28-58](https://user-images.githubusercontent.com/1423607/82972146-5baffd00-9fa2-11ea-95d3-383bd4d3316e.png)

romgrk commented on tests/object__nongi.js:
```diff
@@ -0,0 +1,15 @@
+
+const src = Gst.ElementFactory.make('videotestsrc')
+expect(src.constructor.name, 'GstVideoTestSrc')
+expect(typeof src.addPad, 'function')
```
A few adjustements:
 - filename could be `object__non-introspected.js`, no need to abbreviate here
 - wrap the test in a `describe(testDescription, () => ...)`
 - I don't insist on it but testing with `instanceof` could be more interesting than with string testing
 - it could be nice to test that `src` is also an `instanceof` all the parent classes of `GstVideoTestSrc`

wotzlaff replied:
I added some test with `instanceof`. The problem for non-introspected objects is that we have no access to the corresponding base classes/constructors, do we?

wotzlaff commented:
Thanks for your comments.
I totally agree with your comments.

I didn't notice that my test were failing, too. Hopefully, it will be fixed with the dependencies added in the last commit. At least for bionic it's working. But I'm still waiting for macOS.

romgrk commented:
Right, you know what, actually that test is also failing on master. And I've found which fucker introduced the bug:

![Screenshot from 2020-05-27 04-41-32](https://user-images.githubusercontent.com/1423607/82997418-59ff2d00-9fd4-11ea-87ff-2d6bd9ef3195.png)

Yeah... 😅

So your work is done, thanks for the PR, great work :)
